libs/language-server/src/lib/lsp/import-reference.spec.ts

```diff
@@ -0,0 +1,107 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import path from 'node:path';
+
+import { NodeFileSystem } from 'langium/node';
+
+import {
+  type JayveeServices,
+  createJayveeServices,
+  isJayveeModel,
+} from '../../lib';
+import {
+  expectNoParserAndLexerErrors,
+  parseTestFileInWorkingDir,
+} from '../../test';
+import { type JayveeModel } from '../ast';
+
+describe('References to imported elements', () => {
+  const WORKING_DIR = path.resolve(__dirname, '../../test/assets/');
+  let services: JayveeServices;
+
+  async function parseModel(
+    relativeTestFilePath: string,
+  ): Promise<JayveeModel> {
+    const document = await parseTestFileInWorkingDir(
+      WORKING_DIR,
+      relativeTestFilePath,
+      services,
+    );
+    expectNoParserAndLexerErrors(document);
+
+    const parsedModel = document.parseResult.value;
+    assert(isJayveeModel(parsedModel), 'Test file is not valid Jayvee model');
+    return parsedModel;
+  }
+
+  beforeEach(() => {
+    // Create language services
+    services = createJayveeServices(NodeFileSystem).Jayvee;
+  });
+
+  it('should resolve reference to imported element', async () => {
+    const relativeTestFilePath =
+      'import-published-references/valid-reference-to-imported-element.jv';
+
+    const model = await parseModel(relativeTestFilePath);
+
+    expect(model.pipelines.length).toEqual(1); // file contains one pipeline
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const blocks = model.pipelines[0]!.blocks;
+    expect(blocks.length).toEqual(1); // pipeline contains one block
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const reference = blocks[0]!.type; // of an imported block type
+
+    expect(reference.ref).toBeDefined();
+  });
+
+  it('should resolve reference to transitively imported element', async () => {
```

Comment:
One of two things are happening with these tests:

1. I am too blind to understand what is tested here, to me the code for invalid and valid test cases looks exactly the same and the `expect` calls are also the same... so I fail to understand what is being tested here.
2. The tests are actually all the same except for the input file.

In case 1, I think the test code could be refactored so it is more obvious what is being tested exactly. In case 2, the tests should be fixed.

Reply:
Yes, the tests are pretty similar. The input files should be specific to the test case, though.

For the valid cases, the reference should be followable (`.ref` is **not** `undefined`).
For the invalid cases, the reference should not be followable  (`.ref` **is** `undefined`).

I'll try to refactor using `test.each`, so we just modify the file path for every valid/invalid case. So there will be one test case for valid and one for invalid examples.

Reply:
Ah yeah, then it is a case 1 issue. In my defense, the only difference then is the `Un` in `toBeUndefined` :D. I think refactoring it into one method that just takes a file path and undefined/defined expectation as param would make this much more readable yeah.

Reply:
I agree, hard to spot the difference :)

Reply:
Done in [ad744b7](https://github.com/jvalue/jayvee/pull/564/commits/ad744b78dcef3bb1bf5ac7dd4c8c77a361322343)